File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!subRoot) return true;
        if(!root) return false;
        if(sameTree(subRoot,root))return true;
        return isSubtree(root->left,subRoot)||isSubtree(root->right,subRoot);
    }
    bool sameTree(TreeNode* subRoot,TreeNode* root){
    if(!subRoot && !root)return true;
    if(!root  || !subRoot || subRoot->val!=root->val) return false;
    return sameTree(subRoot->left,root->left) && sameTree(subRoot->right,root->right);
}
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.pathSum=float('-inf')
        def dfs(node):
            if node is None:
                return 0
            left=max(0,dfs(node.left))
            right=max(0,dfs(node.right))
            self.pathSum=max(self.pathSum,left+right+node.val)
            return max(left,right)+node.val
        dfs(root)
        return self.pathSum
        